### PR TITLE
Add more detail error messages when installing with some components has failed.

### PR DESCRIPTION
**Pull request by h-michael**
Error message that we get when installing with some components failed may seem unkind for beginners.
Many users create issue at component's repository.
For example,
https://github.com/rust-lang/rls/issues/1186
https://github.com/rust-lang/rls/issues/641
https://github.com/rust-lang/rustfmt/issues/3244
https://github.com/rust-lang/rust-clippy/issues/2869

So I might want to add more information.
(But there may be more suitable message.)

### CODE REVIEW

File: src/rustup-dist/src/errors.rs [hunk 1]
```diff
@@ -127,11 +127,25 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest) -> String {
         if same_target {
             let mut cs_strs = cs.iter().map(|c| format!("'{}'", c.short_name(manifest)));
             let cs_str = cs_strs.join(", ");
-            let _ = write!(buf, "some components unavailable for download: {}", cs_str);
+            let _ = write!(
+                    buf,
+                    "some components unavailable for download: {}\n\
+                    if you want these components, please install and use latest successful build version, \
```
Comment (kinnison) [inline src/rustup-dist/src/errors.rs]: You're missing a 'the' between 'use' and 'latest'

File: src/rustup-dist/src/errors.rs [hunk 2]
```diff
@@ -127,11 +127,25 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest) -> String {
...
+                    which you can find out at https://rust-lang-nursery.github.io/rust-toolstate, like this.\n\
```
Comment (kinnison) [inline src/rustup-dist/src/errors.rs]: I'd drop the 'out' replace the ',' with '.' and change 'like this.' to '\nTo do this, run something like:\n'

File: src/rustup-dist/src/errors.rs [hunk 3]
```diff
@@ -127,11 +127,25 @@ fn component_unavailable_msg(cs: &[Component], manifest: &Manifest) -> String {
...
+                    rustup install nightly-2018-12-27",
+                    cs_str
+                );
```
Comment (kinnison) [inline src/rustup-dist/src/errors.rs]: Same comments apply here

File: tests/cli-misc.rs [hunk 4]
```diff
@@ -808,7 +808,10 @@ fn update_unavailable_rustc() {
         expect_err(
             config,
             &["rustup", "update", "nightly"],
-            "some components unavailable for download: 'rustc', 'cargo'",
+            "some components unavailable for download: 'rustc', 'cargo'\n\
+            if you want these components, please install and use latest successful build version, \
+            which you can find out at https://rust-lang-nursery.github.io/rust-toolstate, like this.\n\
+            rustup install nightly-2018-12-27",
```
Comment (kinnison) [inline tests/cli-misc.rs]: Same comments apply here.

**Review by kinnison — COMMENTED** [review]
Over-all the concept is probably good, though I worry about directing people to the rust-toolstate page directly.

I wonder if we should consider different messages depending on the channel selected.  If this is nightly, the message you suggest would be suitable, but if this is stable or beta, that suggests something less good going on which might benefit from a different kind of message.

File: src/rustup-dist/src/errors.rs (hunk 1 above)
Comment (nrc) [inline src/rustup-dist/src/errors.rs]:
```suggestion
                    if you require these components, please install and use the latest successful build, \
```

File: src/rustup-dist/src/errors.rs (hunk 2 above)
Comment (nrc) [inline src/rustup-dist/src/errors.rs]:
```suggestion
                    which you can find at https://rust-lang-nursery.github.io/rust-toolstate, for example.\n\
```

File: src/rustup-dist/src/errors.rs (hunk 3 above)
Comment (nrc) [reply src/rustup-dist/src/errors.rs]: Could you pull the message out into a const please

**Review by nrc — APPROVED** [review]
lgtm with the requested changes

**Comment by h-michael** [comment]
@nrc, @kinnison thanks for your review.

> Could you pull the message out into a const please

How can I use write! macro with const literal ?

**Comment by nrc** [comment]
I was thinking you would use something like `write!(buf, "some components unavailable for download: {}\n{}", ..., TOOLSTATE_MSG)` where `TOOLSTATE_MSG` is the const).

**Comment by h-michael** [comment]
@nrc I see, thanks!

**Comment by h-michael** [comment]
@nrc I've fixed.
But test is failed. This may be caused by #1547.

**Comment by h-michael** [comment]
@nrc Fixed and passed CI

**Comment by nrc** [comment]
Thanks!

**Pull request merged**